src/rbindlist.c

```diff
@@ -514,20 +514,24 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
         if (w==-1 || !length(thisCol=VECTOR_ELT(li, w))) {  // !length for zeroCol warning above; #1871
           writeNA(target, ansloc, thisnrow);  // writeNA is integer64 aware and writes INT64_MIN
         } else {
-          if ((TYPEOF(target)==VECSXP || TYPEOF(target)==EXPRSXP) && TYPEOF(thisCol)!=TYPEOF(target)) {
-            // do an as.list() on the atomic column; #3528
-            thisCol = PROTECT(coerceVector(thisCol, TYPEOF(target))); nprotect++;
+          if (maxType==VECSXP) {
+            thisCol = PROTECT(coerceAsList(thisCol, thisnrow)); nprotect++;
           }
```

Comment:
Logic abstracted into a separate function in utils.c because it is otherwise duplicated in asmatrix in #4144. I also envision making an R exposable wrapper so that non-atomic vector columns may be caught and coerced to list in `as.data.table` and `setDT` (future PR once this is merged)

Reply:
Not sure if a single PROTECT is sufficient here - the PROTECT stack in coerceAsList may be `thisnrow + 1` if `TYPEOF(thisCol)==EXPRSXP` because each element in the vector has to be coerced to an EXPRSXP and passed to SET_VECTOR_ELT in the list wrapper, otherwise the list elements are `LANGSXP` if you just do `coerceVector(thisCol, VECSXP)`. There is a corresponding UNPROTECT call in `coerceAsList` that handles this, but I don't know if that means wrapping the result here in a single PROTECT risks this memory being freed before rbindlist returns.

Reply:
The method i've settled on after hitting the protect stack limit in testing #4144 is to have the C function called from R manage the PROTECT stack, and for functions it calls to simply increment a protect counter. 

I've modified `coerceAsList` to this effect in the latest commit. Inside `coerceAsList`, each new allocated list element PROTECT is popped from the stack once it's been assigned into the new list using SET_VECTOR_ELT. In the coercion loop in rbindlist.c I make a separate nprotect counter for `coerceAsList`, as it may either increment the PROTECT stack by one, or not at all (e.g. when simply returning the input list). This allows us to pop the coerced list from the PROTECT stack after its contents have been copied into the target vector. This means these changes will no longer hit the PROTECT stack limit when handling >50,000 columns that need to be converted to lists.